<center>
    <img src="https://gitlab.com/ibm/skills-network/courses/placeholder101/-/raw/master/labs/module%201/images/IDSNlogo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>


# **Launch Sites Locations Analysis with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [1]:
!pip3 install --force-reinstal folium
!pip3 install wget

  Using cached folium-0.12.1.post1-py2.py3-none-any.whl (95 kB)
  Using cached numpy-1.21.5-cp37-cp37m-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (15.7 MB)
  Using cached Jinja2-3.0.3-py3-none-any.whl (133 kB)
  Using cached requests-2.26.0-py2.py3-none-any.whl (62 kB)
  Using cached branca-0.4.2-py3-none-any.whl (24 kB)
  Using cached MarkupSafe-2.0.1-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (31 kB)
  Using cached certifi-2021.10.8-py2.py3-none-any.whl (149 kB)
  Using cached urllib3-1.26.7-py2.py3-none-any.whl (138 kB)
  Using cached idna-3.3-py3-none-any.whl (61 kB)
  Using cached charset_normalizer-2.0.9-py3-none-any.whl (39 kB)
  Attempting uninstall: MarkupSafe
    Found existing installation: MarkupSafe 2.0.1
    Uninstalling MarkupSafe-2.0.1:
      Successfully uninstalled MarkupSafe-2.0.1
  Attempting uninstall: urllib3
    Found existing installation: urllib3 1.26.7
    Uninstalling urllib3-1.26.7:
      Successfully

In [2]:
import folium
import wget
import pandas as pd

In [3]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module\_3/DV0101EN-3-5-1-Generating-Maps-in-Python-py-v2.0.ipynb)


## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [4]:
# Download and read the `spacex_launch_geo.csv`
spacex_csv_file = wget.download('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv')
spacex_df=pd.read_csv(spacex_csv_file)

Now, you can take a look at what are the coordinates for each site.


In [5]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610746

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [6]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [7]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [8]:
# Initial the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=5, control_scale=True, zoomDelta=0.05, zoomSnap=0.05)
# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. In addition, add Launch site name as a popup label
CCAFS_LC_40 = [28.56230196799018, -80.57735647504778]
circle = folium.Circle(CCAFS_LC_40, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('CCAFS LC-40'))
marker = folium.map.Marker(
    CCAFS_LC_40,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'CCAFS LC-40',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)


CCAFS_SLC_40 = [28.563197177407144, -80.57682003124195]
circle = folium.Circle(CCAFS_SLC_40, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('CCAFS SLC-40'))
marker = folium.map.Marker(
    CCAFS_SLC_40,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'CCAFS SLC-40',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)


KSC_LC_39A = [28.57325457037815, -80.64689528960382]
circle = folium.Circle(KSC_LC_39A, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('KSC LC-39A'))
marker = folium.map.Marker(
    KSC_LC_39A,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'KSC LC-39A',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)


VAFB_SLC_4E = [34.632834161782775, -120.61074553068327]
circle = folium.Circle(VAFB_SLC_4E, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('VAFB SLC-4E'))
marker = folium.map.Marker(
    VAFB_SLC_4E,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'VAFB SLC-4E',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png" />
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?
*   Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


# Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [9]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [10]:
marker_cluster = MarkerCluster()

*TODO:* Create a new column in `launch_sites` dataframe called `marker_color` to store the marker colors based on the `class` value


In [11]:

# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red


In [12]:
# Function to assign color to launch outcome
def assign_marker_color(launch_outcome):
    if launch_outcome == 1:
        return 'green'
    else:
        return 'red'
    
spacex_df['marker_color'] = spacex_df['class'].apply(assign_marker_color)
spacex_df.tail(10)

Launch Site        Lat       Long  class marker_color
46    KSC LC-39A  28.573255 -80.646895      1        green
47    KSC LC-39A  28.573255 -80.646895      1        green
48    KSC LC-39A  28.573255 -80.646895      1        green
49  CCAFS SLC-40  28.563197 -80.576820      1        green
50  CCAFS SLC-40  28.563197 -80.576820      1        green
51  CCAFS SLC-40  28.563197 -80.576820      0          red
52  CCAFS SLC-40  28.563197 -80.576820      0          red
53  CCAFS SLC-40  28.563197 -80.576820      0          red
54  CCAFS SLC-40  28.563197 -80.576820      1        green
55  CCAFS SLC-40  28.563197 -80.576820      0          red

*TODO:* For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [13]:
# Add marker_cluster to current site_map
site_map.add_child(marker_cluster)

# for each row in spacex_df data frame
# create a Marker object with its coordinate
# and customize the Marker's icon property to indicate if this launch was successed or failed, 
# e.g., icon=folium.Icon(color='white', icon_color=row['marker_color']
for index, record in spacex_df.iterrows():
    # TODO: Create and add a Marker cluster to the site map
    # marker = folium.Marker(...)
    lat=record['Lat']
    lng=record['Long']
    label=record['Launch Site']
    color=record['marker_color']
    marker = folium.Marker(
        radius=10,
        location=[lat, lng],
        icon=folium.Icon(color='white', icon_color=color, icon='info-sign'),
        fill=True,
        popup=label,
        fill_color=color,    
    )
    marker_cluster.add_child(marker)

site_map

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png" />
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png" />
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


# TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [14]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


You can calculate the distance between two points on the map based on their `Lat` and `Long` values using the following method:


In [15]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

*TODO:* Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [16]:
# find coordinate of the closet coastline
# e.g.,: Lat: 28.56367  Lon: -80.57163
# distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)
distance_coastline = calculate_distance(28.563197177407144, -80.57682003124195, 28.56312, -80.56794)

In [17]:
distance_coastline

0.8675515178052864

*TODO:* After obtained its coordinate, create a `folium.Marker` to show the distance


In [18]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
# distance_marker = folium.Marker(
#    coordinate,
#    icon=DivIcon(
#        icon_size=(20,20),
#        icon_anchor=(0,0),
#        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance),
#        )
#    )

coordinate=[28.56312, -80.56794]

distance_marker = folium.Marker(
   coordinate,
   icon=DivIcon(
       icon_size=(20,20),
       icon_anchor=(0,0),
       html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(0.8675515178052864),
       )
   )

site_map.add_child(distance_marker)
site_map

*TODO:* Draw a `PolyLine` between a launch site to the selected coastline point


In [19]:
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
coordinates=[[28.563197177407144, -80.57682003124195], [28.56312, -80.56794]]
lines=folium.PolyLine(locations=coordinates, weight=1)
site_map.add_child(lines)

Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png" />
</center>


*TODO:* Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png" />
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png" />
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png" />
</center>


Closest Railroad to KSC LC-39A, CCAFS LC-40, and CCAFC SLC-40

In [20]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site
distance_railways = calculate_distance(28.563197177407144, -80.57682003124195, 28.57206, -80.58527)
distance_railways
railways=[28.57206, -80.58527]

distance_marker = folium.Marker(
   railways,
   icon=DivIcon(
       icon_size=(20,20),
       icon_anchor=(0,0),
       html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_railways),
       )
   )

site_map.add_child(distance_marker)
coordinates2=[[28.563197177407144, -80.57682003124195], [28.57206, -80.58527]]
lines=folium.PolyLine(locations=coordinates2, weight=1)
site_map.add_child(lines)

Closest Railroad to VAFB SLC-4E

In [21]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site
distance_railways2 = calculate_distance(34.632834161782775, -120.61074553068327, 34.6351, -120.62418)
distance_railways2
railways2=[34.6351, -120.62418]

distance_marker = folium.Marker(
   railways2,
   icon=DivIcon(
       icon_size=(20,20),
       icon_anchor=(0,0),
       html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_railways2),
       )
   )

site_map.add_child(distance_marker)
coordinates3=[[34.632834161782775, -120.61074553068327], [34.6351, -120.62418]]
lines=folium.PolyLine(locations=coordinates3, weight=1)
site_map.add_child(lines)

Closest Highway to KSC LC-39A, CCAFS LC-40, and CCAFC SLC-40

In [22]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site
distance_highways = calculate_distance(28.57325457037815, -80.64689528960382, 28.57325457037815, -80.65548)
distance_highways
highways=[28.57325457037815, -80.65548]

distance_marker = folium.Marker(
   highways,
   icon=DivIcon(
       icon_size=(20,20),
       icon_anchor=(0,0),
       html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_highways),
       )
   )

site_map.add_child(distance_marker)
coordinates2=[[28.57325457037815, -80.64689528960382], [28.57325457037815, -80.65548]]
lines=folium.PolyLine(locations=coordinates2, weight=1)
site_map.add_child(lines)

Closest City to VAFB SLC-4E

In [23]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site
distance_city = calculate_distance(34.632834161782775, -120.61074553068327, 34.64364, -120.45779)
distance_city
city=[34.64364, -120.45779]

distance_marker = folium.Marker(
   city,
   icon=DivIcon(
       icon_size=(20,20),
       icon_anchor=(0,0),
       html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_city),
       )
   )

site_map.add_child(distance_marker)
coordinates3=[[34.632834161782775, -120.61074553068327], [34.64364, -120.45779]]
lines=folium.PolyLine(locations=coordinates3, weight=1)
site_map.add_child(lines)

Airports (purple circles) close to KSC LC-39A, CCAFS LC-40, and CCAFC SLC-40
* blue circle 20 km range
* green circle 50 km range

In [24]:
ring1 = folium.Circle([28.5662512385918,-80.6003572652978], radius=20000, color='#1053DA', fill=False)
ring2 = folium.Circle([28.5662512385918,-80.6003572652978], radius=50000, color='#009900', fill=False)
circle01 = folium.Circle([28.613, -80.694], radius=250, color='#990099', fill=True)
circle02 = folium.Circle([28.465, -80.561], radius=250, color='#990099', fill=True)
circle03 = folium.Circle([28.862, -80.911], radius=250, color='#990099', fill=True)
circle04 = folium.Circle([28.886, -80.866], radius=250, color='#990099', fill=True)
circle05 = folium.Circle([28.760, -80.849], radius=250, color='#990099', fill=True)
circle06 = folium.Circle([28.666, -81.083], radius=250, color='#990099', fill=True)
circle07 = folium.Circle([28.622, -80.834], radius=250, color='#990099', fill=True)
circle08 = folium.Circle([28.513, -80.799], radius=250, color='#990099', fill=True)
circle09 = folium.Circle([28.298, -80.720], radius=250, color='#990099', fill=True)
circle10 = folium.Circle([28.341, -80.688], radius=250, color='#990099', fill=True)
circle11 = folium.Circle([28.235, -80.610], radius=250, color='#990099', fill=True)
site_map.add_child(ring1)
site_map.add_child(ring2)
site_map.add_child(circle01)
site_map.add_child(circle02)
site_map.add_child(circle03)
site_map.add_child(circle04)
site_map.add_child(circle05)
site_map.add_child(circle06)
site_map.add_child(circle07)
site_map.add_child(circle08)
site_map.add_child(circle09)
site_map.add_child(circle10)
site_map.add_child(circle11)

Airports (purple circles) close to VAB SLC-4E
* blue circle 20 km range
* green circle 50 km range

In [25]:
ring1 = folium.Circle([34.632834161782775, -120.61074553068327], radius=20000, color='#1053DA', fill=False)
ring2 = folium.Circle([34.632834161782775, -120.61074553068327], radius=50000, color='#009900', fill=False)
circle01 = folium.Circle([34.737, -120.582], radius=250, color='#990099', fill=True)
circle02 = folium.Circle([34.665, -120.468], radius=250, color='#990099', fill=True)
circle03 = folium.Circle([34.607, -120.076], radius=250, color='#990099', fill=True)
circle04 = folium.Circle([34.899, -120.454], radius=250, color='#990099', fill=True)
circle05 = folium.Circle([34.988, -120.459], radius=250, color='#990099', fill=True)
site_map.add_child(ring1)
site_map.add_child(ring2)
site_map.add_child(circle01)
site_map.add_child(circle02)
site_map.add_child(circle03)
site_map.add_child(circle04)
site_map.add_child(circle05)

After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways?
*   Are launch sites in close proximity to highways?
*   Are launch sites in close proximity to coastline?
*   Do launch sites keep certain distance away from cities?

Also please try to explain your findings.


# <font color='green'>**My Findings:**</font>
* <font color='green'>**the closest coastline – 0.87 km from CAFS LC-40/CAFC SLC-40**</font>
* <font color='green'>**the closest railway – 1.29 km from CAFS LC-40/CAFC SLC-40 and 1.26 km from VAFB SLC-4E**</font>
* <font color='green'>**the closest highway – 0.84 km from KSC LC-39**</font>
* <font color='green'>**the closest city – 14.05 km from VAFB SLC-4E**</font>
* <font color='green'>**additionally, there is much higher number of airports in proximity to CAFS LC-40/CAFC SLC-40/ KSC LC-39 than to VAFB SLC-4E**</font>

# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Yan Luo](https://www.linkedin.com/in/yan-luo-96288783/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2021-01-01)


### Other Contributors


Joseph Santarcangelo


## Change Log


| Date (YYYY-MM-DD) | Version | Changed By | Change Description          |
| ----------------- | ------- | ---------- | --------------------------- |
| 2021-05-26        | 1.0     | Yan        | Created the initial version |


Copyright © 2021 IBM Corporation. All rights reserved.
